src/circuit.rs

```diff
@@ -2,61 +2,771 @@
 
 use std::mem;
 
+use group::{Curve, GroupEncoding};
 use halo2::{
     circuit::{Layouter, SimpleFloorPlanner},
-    plonk,
+    plonk::{self, Advice, Column, Expression, Instance as InstanceColumn, Selector},
     poly::Rotation,
     transcript::{Blake2bRead, Blake2bWrite},
 };
-use pasta_curves::{pallas, vesta};
+use pasta_curves::{
+    arithmetic::{CurveAffine, FieldExt},
+    pallas, vesta,
+};
 
 use crate::{
-    note::{nullifier::Nullifier, ExtractedNoteCommitment},
-    primitives::redpallas::{SpendAuth, VerificationKey},
+    constants::{
+        load::{NullifierK, OrchardFixedBasesFull, ValueCommitV},
+        MERKLE_DEPTH_ORCHARD,
+    },
+    keys::{
+        CommitIvkRandomness, DiversifiedTransmissionKey, NullifierDerivingKey, SpendValidatingKey,
+    },
+    note::{
+        commitment::{NoteCommitTrapdoor, NoteCommitment},
+        nullifier::Nullifier,
+        ExtractedNoteCommitment,
+    },
+    primitives::{
+        poseidon::{self, ConstantLength},
+        redpallas::{SpendAuth, VerificationKey},
+    },
+    spec::NonIdentityPallasPoint,
     tree::Anchor,
-    value::ValueCommitment,
+    value::{NoteValue, ValueCommitTrapdoor, ValueCommitment},
+};
+use gadget::{
+    ecc::{
+        chip::{EccChip, EccConfig},
+        FixedPoint, FixedPointBaseField, FixedPointShort, Point,
+    },
+    poseidon::{
+        Hash as PoseidonHash, Pow5T3Chip as PoseidonChip, Pow5T3Config as PoseidonConfig,
+        StateWord, Word,
+    },
+    sinsemilla::{
+        chip::{SinsemillaChip, SinsemillaConfig, SinsemillaHashDomains},
+        commit_ivk::CommitIvkConfig,
+        merkle::{
+            chip::{MerkleChip, MerkleConfig},
+            MerklePath,
+        },
+        note_commit::NoteCommitConfig,
+    },
+    utilities::{
+        copy,
+        enable_flag::{EnableFlagChip, EnableFlagConfig},
+        plonk::{PLONKChip, PLONKConfig, PLONKInstructions},
+        CellValue, UtilitiesInstructions, Var,
+    },
 };
 
+use std::convert::TryInto;
+
 pub(crate) mod gadget;
 
 /// Size of the Orchard circuit.
-const K: u32 = 11;
+// FIXME: This circuit should fit within 2^11 rows.
+const K: u32 = 12;
+
+// Absolute offsets for public inputs.
+const ZERO: usize = 0;
+const ANCHOR: usize = 1;
+const NF_OLD: usize = 2;
+const CV_NET_X: usize = 3;
+const CV_NET_Y: usize = 4;
+const RK_X: usize = 5;
+const RK_Y: usize = 6;
+const CMX: usize = 7;
+const ENABLE_SPEND: usize = 8;
+const ENABLE_OUTPUT: usize = 9;
```

Comment:
This doesn't match the spec.
![image](https://user-images.githubusercontent.com/643204/126200257-c6663dc6-cc07-4bb1-8b95-c67bc5942d9d.png)

Reply:
Hm, it's much simpler (implementation-wise) to bring `enableOutputs` and `enableSpends` in individually as booleans. What is the motivation for packing them together like this?

Reply:
For Groth16, inputs were quite expensive (they required a scalar multiplication each, for unbatched verification). I don't know to what extent that is still the case for Halo 2. @ebfull ?

Reply:
If the cost is not significant, I can split enableSpends and enableOutputs into separate field elements. Note that that isn't the only difference between the implementation above and the spec. When the constants API is added, there will be no need for a zero instance cell.